metals/src/main/scala/scala/meta/internal/builds/Digest.scala

```diff
@@ -0,0 +1,251 @@
+  ): Boolean = {
+    try {
+      Files
+        .readAllLines(file.toNIO)
+        .asScala
+        .mkString("\n")
+        .replaceAll("""/\*([^*]|[\r\n]|(\*+([^*/]|[\r\n])))*\*+/""", "") // match anything between /**/ comments and remove
+        .replaceAll("""//.*""", "") // replace any inline comment
+        .split("\\s+")
+        .foreach { word =>
+          digest.update(word.getBytes())
+        }
+      true
+    } catch {
+      case NonFatal(_) =>
+        false
+    }
+  }
+
+  def digestScala(
+      file: AbsolutePath,
+      digest: MessageDigest
+  ): Boolean = {
+    try {
+      val input = file.toInput
+      val tokens = input.tokenize.get
+      tokens.foreach {
+        case _: Token.Space | _: Token.Tab | _: Token.CR | _: Token.LF |
+            _: Token.LFLF | _: Token.FF | _: Token.Comment | _: Token.BOF |
+            _: Token.EOF => // Do nothing
+        case token =>
+          val bytes = StandardCharsets.UTF_8.encode(token.pos.text)
+          digest.update(token.productPrefix.getBytes())
+          digest.update(bytes)
+      }
+      true
+    } catch {
+      case NonFatal(e) =>
+        false
+    }
+  }
+}
+
+trait Digestable {
+  def current(workspace: AbsolutePath): Option[String] = {
+    if (!workspace.isDirectory) None
+    else {
+      val digest = MessageDigest.getInstance("MD5")
+      // we skip the version in tests, so that we don't have to manually update the digests in tests
+      // when changing the version
+      if (System.getProperty("metals.testing") == null) {
+        digest.update(Digest.version.getBytes(StandardCharsets.UTF_8))
+      }
+
+      val isSuccess = digestWorkspace(workspace, digest)
+      if (isSuccess) Some(MD5.bytesToHex(digest.digest()))
+      else None
+    }
+  }
+
+  protected def digestWorkspace(
+      absolutePath: AbsolutePath,
+      digest: MessageDigest
+  ): Boolean
+}
+
+object SbtDigest extends Digestable {
+  override protected def digestWorkspace(
+      workspace: AbsolutePath,
+      digest: MessageDigest
+  ): Boolean = {
+    val project = workspace.resolve("project")
+    Digest.digestDirectory(workspace, digest) &&
+    Digest.digestFileBytes(project.resolve("build.properties"), digest) &&
+    Digest.digestDirectory(project, digest) &&
+    Digest.digestDirectory(project.resolve("project"), digest)
+  }
+}
+
+object GradleDigest extends Digestable {
+  override protected def digestWorkspace(
+      workspace: AbsolutePath,
+      digest: MessageDigest
+  ): Boolean = {
+    val buildSrc = workspace.resolve("buildSrc")
+    val buildSrcDigest = if (buildSrc.isDirectory) {
+      digestBuildSrc(buildSrc, digest)
+    } else {
+      true
+    }
+    buildSrcDigest && Digest.digestDirectory(workspace, digest) && digestSubProjects(
+      workspace,
+      digest
+    )
+  }
+
+  def digestBuildSrc(path: AbsolutePath, digest: MessageDigest): Boolean = {
+    var isSuccessful = true
+    Files.walk(path.toNIO).forEach { file =>
```

Comment:
Second time - maybe it could be extracted  as some kind of `def fileIterator(path): Iterator[AbsolutePath]` and then we could use `fileIterator(path).forAll(file => digestFile(file, digest))`